drivers/flash/flash_mcux_xspi.c

```diff
@@ -0,0 +1,490 @@
+	},
+
+	/* Page program. */
+	[FLASH_CMD_PAGEPROGRAM_OCTAL] = {
+		XSPI_LUT_SEQ(kXSPI_Command_DDR, kXSPI_8PAD, 0x12, kXSPI_Command_DDR, kXSPI_8PAD, 0xED),
+		XSPI_LUT_SEQ(kXSPI_Command_RADDR_DDR, kXSPI_8PAD, 0x20, kXSPI_Command_WRITE_DDR, kXSPI_8PAD, 0x8),
+	}
+};
+
+static int flash_xspi_nor_wait_bus_busy(const struct device *dev, bool enableOctal)
+{
+	struct flash_mcux_xspi_data *devData = dev->data;
+	const struct device *xspi_dev = devData->xspi_dev;
+	xspi_transfer_t flashXfer;
+	uint32_t readValue;
+	bool isBusy;
+	int ret;
+
+	flashXfer.deviceAddress = devData->amba_address;
+	flashXfer.cmdType       = kXSPI_Read;
+	flashXfer.data          = &readValue;
+	flashXfer.targetGroup   = kXSPI_TargetGroup0;
+	flashXfer.dataSize      = enableOctal ? 2 : 1;
+	flashXfer.seqIndex      = enableOctal ? FLASH_CMD_READ_STATUS_OPI : FLASH_CMD_READ_STATUS;
+	flashXfer.lockArbitration = false;
+
+	do {
+		ret = memc_mcux_xspi_transfer(xspi_dev, &flashXfer);
+		if (ret < 0)
+		{
+			return ret;
+		}
+
+		isBusy = (readValue & (1U << FLASH_BUSY_STATUS_OFFSET)) ? true : false;
+	} while (isBusy);
+
+	return 0;
+}
+
+static int flash_mcux_xspi_read(const struct device *dev, off_t offset, void *data, size_t len)
+{
+	struct flash_mcux_xspi_data *devData = dev->data;
+	const struct device *xspi_dev = devData->xspi_dev;
+	xspi_transfer_t flashXfer;
+	int ret;
+
+	flashXfer.deviceAddress   = devData->amba_address + offset;
+	flashXfer.cmdType         = kXSPI_Read;
+	flashXfer.seqIndex        = FLASH_CMD_MEM_READ;
+	flashXfer.targetGroup     = kXSPI_TargetGroup0;
+	flashXfer.data            = (uint32_t *)data;
+	flashXfer.dataSize        = len;
+	flashXfer.lockArbitration = false;
+
+	ret = memc_mcux_xspi_transfer(xspi_dev, &flashXfer);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return flash_xspi_nor_wait_bus_busy(dev, true);
+}
+
+static int flash_mcux_xspi_write_enable(const struct device *dev, uint32_t baseAddr, bool enableOctal)
+{
+	struct flash_mcux_xspi_data *data = dev->data;
+	const struct device *xspi_dev = data->xspi_dev;
+	xspi_transfer_t flashXfer;
+
+	flashXfer.deviceAddress   = data->amba_address + baseAddr;
+	flashXfer.cmdType         = kXSPI_Command;
+	flashXfer.targetGroup     = kXSPI_TargetGroup0;
+	flashXfer.data            = NULL;
+	flashXfer.dataSize        = 0UL;
+	flashXfer.lockArbitration = false;
+	flashXfer.seqIndex        = enableOctal ? FLASH_CMD_WRITE_ENABLE_OPI : FLASH_CMD_WRITE_ENABLE;
+
+	return memc_mcux_xspi_transfer(xspi_dev, &flashXfer);
+}
+
+static int flash_mcux_xspi_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	struct flash_mcux_xspi_data *devData = dev->data;
+	const struct device *xspi_dev = devData->xspi_dev;
+	xspi_transfer_t flashXfer;
+	uint32_t key = 0;
+	int ret;
+
+	if (memc_xspi_is_running_xip(xspi_dev)) {
+		key = irq_lock();
+		memc_xspi_wait_bus_idle(xspi_dev);
+	}
+
+	ret = flash_mcux_xspi_write_enable(dev, 0, true);
+	if (ret < 0) {
+		return ret;
```

Comment:
this may return with irq locked, should unlock irq if `irq_lock` is called in this function. Please also check other `return` code in this function.